### ISSUE TITLE
Feature Request - Allow retry until with condition of NoChanges

### ISSUE THREAD

**Issue by crux-capacitor** [saltstack/salt#51783]
### Description of Issue/Question

I would like to suggest allowing the 'retry' requisite to have an 'until' condition of 'NoChanges'. The idea is that the state will be retried until there are no changes.

### Use Case:

I'm working on a formula to install and configure NiFi. During the course of the entire run, a basic config file (nifi.properties) is managed. Prior to the file being managed, a TLS toolkit script gets run which generates a keystore, truststore and some passwords stored in another file (tls_value_file). 

Those values need to be put into the nifi.properties file and are gathered from that file using jinja commands wrapped in an if-block , however due to the execution order of jinja versus states (all jinja being rendered first), when the jinja gets evaluated, the file containing those values doesn't exist yet. 

In the config.sls which manages the nifi.properties file, I have an if-block (if salt.file.file_exists(tls_value_file)) that gathers those values if true. Since that file doesn't exist during the first run of the config.sls, the values end up empty. I want to be able to apply a retry onto the file.manged with the until condition of 'NoChanges' so that it runs until there are no changes.

I hope I explained my use case coherently! :)

**Comment by dwoz** [comment]
@saltstack/team-core Thoughts on this?

**Comment by waynew** [comment]
Hey @crux-capacitor, thanks for the feature request!

I'm personally a fan of the idea - though I can think of at least one potential problem off the top of my head. Consider a state where you're bringing up a service, but there's an error in the config file (or worse, an error that takes a second or two to crash the service). When it tries to start there will be changes. Then it will try again, and continue trying until the heat death of the universe or your server :trollface: 

That being said, it makes sense that you could have like a `max_tries` or something to limit that.

There may be other ramifications that I'm not thinking about/am aware of, but that's at least one that would need to be considered/fleshed out at least.

But generally speaking, I'm kind of a fan of being able to say, "Just go at it until everything is configured and setup the way it should be." Like another use case would be something like adding an external pillar source - if you make a change and add that, and then that change requires changes, it would be kind of cool to fire and forget.

I did watch an interesting talk a couple of years back where someone just had a mulit-master setup where they were configured to build themselves and then rebuild the other master if it was missing/failed. I don't remember the exact mechanics, but it's possible that you could do something like this with reactors already? Not sure.

**Comment by crux-capacitor** [comment]
@waynew Hi Wayne! I'm glad you like the idea. I definitely agree about enforcing a max attempts for this. The existing retry requisite has an attempts argument already, which defaults to 2. So adding a new until condition to the same requisite would make it subject to the default of 2 attempts.

One consideration to take into place - in the use case that I made where a file gets created during the state run and then values are being taken from that file and injected into a managed file - the entire sls with that code would need to be reloaded / re-rendered on each iteration of the retry so that the jinja variables would be refreshed, so to speak.

As far as other pitfalls, you would definitely know better than me what could potentially go wrong :)

If this doesn't get traction, though, I was able to workaround it by putting my jinja code into the managed file itself rather than the state.
